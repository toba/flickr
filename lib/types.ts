export namespace Flickr {
   enum Boolean {
      'false',
      'true'
   }

   export enum SizeUrl {
      Thumb = 'url_t',
      Square75 = 'url_sq',
      Square150 = 'url_q',
      Small240 = 'url_s',
      Small320 = 'url_n',
      Medium500 = 'url_m',
      Medium640 = 'url_z',
      Medium800 = 'url_c',
      Large1024 = 'url_l',
      Large1600 = 'url_h',
      Large2048 = 'url_k',
      Original = 'url_o'
   }

   export enum TypeName {
      User = 'user_id',
      Set = 'photoset_id',
      Photo = 'photo_id'
   }

   export enum Status {
      Okay = 'ok',
      Failed = 'fail'
   }

   export enum Format {
      /** https://www.flickr.com/services/api/response.json.html */
      JSON = 'json',
      /** https://www.flickr.com/services/api/response.xmlrpc.html */
      XML = 'xml'
   }

   export enum Sort {
      DatePostedAsc = 'date-posted-asc',
      DatePostedDesc = 'date-posted-desc',
      DateTakenAsc = 'date-taken-asc',
      DateTakenDesc = 'date-taken-desc',
      InterestingDesc = 'interestingness-desc',
      InterestingAsc = 'interestingness-asc',
      Relevance = 'relevance'
   }

   /**
    * http://www.flickr.com/services/api/flickr.photos.licenses.getInfo.html
    */
   enum License {
      AllRightsReserved = 0,
      Attribution = 4,
      Attribution_NoDervis = 6,
      Attribution_NonCommercial_NoDerivs = 3,
      Attribution_NonCommercial = 2,
      Attribution_NonCommercial_ShareAlike = 1,
      Attribution_ShareAlike = 4,
      NoKnownRestriction = 7,
      UnitedStatesGovernmentWork = 8
   }
   /**
    * http://www.flickr.com/services/api/flickr.photos.setSafetyLevel.html
    */
   enum SafetyLevel {
      Safe = 1,
      Moderate = 2,
      Restricted = 3
   }

   export enum Extra {
      Description = 'description',
      Tags = 'tags',
      DateTaken = 'date_taken',
      Location = 'geo',
      PathAlias = 'path_alias'
   }

   /**
    * Parameters required or allowed with a Flickr API request.
    * https://www.flickr.com/services/api/flickr.photos.search.html
    */
   export interface Params {
      [index: string]: string | number | boolean | string[];
      /** https://www.flickr.com/services/api/misc.api_keys.html */
      api_key?: string;
      format?: Format;
      nojsoncallback?: Boolean;
      method?: string;
      /** Comma-delimited list of method-specific, extra fields to return */
      extras?: string;
      tags?: string;
      sort?: Sort;
      /**
       * Numer of items to return per page of results. The maximum is 500.
       */
      per_page?: number;
      [TypeName.Photo]?: string;
      [TypeName.User]?: string;
      [TypeName.Set]?: string;
   }

   export interface Collection {
      id: string;
      title: string;
      description: string;
      iconlarge: string;
      iconsmall: string;
      collection: Collection[];
      set: SetSummary[];
   }

   export interface Content {
      _content: string;
   }

   export interface EditAbility {
      cancomment: Boolean;
      canaddmeta: Boolean;
   }

   export interface Exif {
      tagspace: string;
      tagspaceid: number;
      tag: string;
      label: string;
      raw: Content;
   }

   interface FarmLocation {
      id: string;
      secret: string;
      server: string;
      farm: number;
   }

   export interface Location {
      latitude: number;
      longitude: number;
      accuracy: number;
      context: number;
      county: Place;
      region: Place;
      country: Place;
   }

   interface LocationPermission {
      ispublic: Boolean;
      iscontent: Boolean;
      isfriend: Boolean;
      isfamily: Boolean;
   }

   export interface MemberSet extends FarmLocation {
      title: string;
      primary: string;
      view_count: number;
      comment_count: number;
      count_photo: number;
      count_video: number;
   }

   interface Owner {
      nsid: string;
      username: string;
      location: string;
      iconserver: string;
      iconfarm: number;
   }

   interface Permission {
      permcomment: Boolean;
      permmetadata: Boolean;
   }

   interface PhotoDates {
      posted: string;
      taken: string;
      takengranularity: number;
      lastupdate: string;
   }

   export interface PhotoInfo extends FarmLocation {
      dateuploaded: string;
      isfavorite: Boolean;
      license: License;
      safetylevel: SafetyLevel;
      rotate: Boolean;
      originalsecret: string;
      originalformat: string;
      owner: Owner;
      title: Content;
      description: Content;
      visibility: Visibility;
      dates: PhotoDates;
      views: number;
      permissions: Permission;
      editability: EditAbility;
      publiceditability: EditAbility;
      usage: Usage;
      EXIF: Exif[];
      tags: {
         tag: TagSummary;
      };
      location: Location;
      geoperms: LocationPermission;
      media: string;
      urls: {
         url: URL[];
      };
   }

   // https://www.flickr.com/services/api/flickr.photos.getExif.html
   // interface PhotoExif {
   //    photo: Flickr.PhotoSummary;
   // }

   export interface PhotoMembership {
      set: MemberSet[];
   }

   export interface SetPhotos {
      photoset: SetInfo;
      photo: PhotoSummary[];
      page: number;
      per_page: string;
      perpage: string;
      pages: number;
      primary: string;
      owner: string;
      ownername: string;
      title: string;
      total: number;
   }

   export interface PhotoSummary extends Place {
      id: string;
      secret: string;
      server: string;
      farm: number;
      title: string;
      isprimary: string;
      tags: string;
      description?: Content;
      datetaken?: string;
      datetakengranularity?: string;
      latitude?: string;
      longitude?: string;
      context?: number;
      geo_is_family?: Boolean | boolean;
      geo_is_friend?: Boolean | boolean;
      geo_is_contact?: Boolean | boolean;
      geo_is_public?: Boolean | boolean;
      lastupdate?: string;
      pathalias?: string;

      exif: Exif[];

      url_s?: string;
      height_s?: string;
      width_s?: string;

      url_h?: string;
      height_h?: string;
      width_h?: string;

      url_k?: string;
      height_k?: string;
      width_k?: string;

      url_l?: string;
      height_l?: string;
      width_l?: string;

      url_m?: string;
      height_m?: string;
      width_m?: string;

      url_o?: string;
      height_o?: string;
      width_o?: string;
   }

   interface Place {
      place_id: string;
      woeid: string;
   }

   export interface Response {
      photoset?: SetPhotos | SetInfo;
      set?: MemberSet[];
      collections?: Tree;
      photo?: PhotoInfo;
      sizes?: SizeList;
      /**
       * Dynamically added property indicating whether the request can be
       * retried.
       */
      retry?: boolean;
      /** Response status */
      stat: Status;
      code?: number;
      message?: string;
      photos?: {
         photo: SearchResult | PhotoSummary[];
      };
      who?: {
         tags: {
            tag: Tag[];
         };
      };
   }

   interface SearchResult {
      page: number;
      pages: number;
      perpage: number;
      total: number;
   }

   export interface SetInfo {
      title: Content;
      description: Content;
      id: string;
      owner: string;
      username: string;
      primary: string;
      secret: string;
      server: string;
      farm: number;
      photos: number;
      count_views: number;
      count_comments: number;
      count_photos: number;
      count_vidoes: number;
      can_comment: Boolean;
      date_create: number;
      date_update: number;
   }

   interface SetSummary {
      id: string;
      title: string;
      description: string;
   }

   export interface Size {
      label: string;
      width: number;
      height: number;
      source: string;
      url: string;
      media: string;
   }

   interface SizeList {
      size: Flickr.Size[];
   }

   export interface Tag {
      clean: string;
      raw: Content[];
   }

   interface TagSummary {
      id: string;
      author: string;
      raw: string;
      machine_tag: number;
   }

   interface Tree {
      collection: Flickr.Collection[];
   }

   interface URL extends Flickr.Content {
      type: string;
   }

   interface Usage {
      candownload: Boolean;
      canblog: Boolean;
      canprint: Boolean;
      canshare: Boolean;
   }

   interface Visibility {
      ispublic: Boolean;
      isfriend: Boolean;
      isfamily: Boolean;
   }

   // enum ExifTag {
   //    Description: 'ImageDescription',
   // CameraMake: 'Make',
   // CameraModel: 'Model',
   // CameraSerialNumber: 'SerialNumber',
   // Lens: 'Lens',
   // LensInfo: 'LensInfo',
   // LensModel: 'LensModel',
   // ResolutionX: 'XResolution',
   // ResolutionY: 'YResolution',
   // ResolutionUnit: 'ResolutionUnit',
   // DisplayedUnitsX: 'DisplayedUnitsX',
   // DisplayedUnitsY: 'DisplayedUnitsY',
   // Software: 'Software',
   // ApplicationRecordVersion: 'ApplicationRecordVersion',
   // DateCreated: 'CreateDate',
   // DateModified: 'ModifyDate',
   // DateOriginal: 'DateTimeOriginal',
   // TimeCreated: 'TimeCreated',
   // MetadataDate: 'MetadataDate',
   // DigitalCreationDate: 'DigitalCreationDate',
   // DigitalCreationTime: 'DigitalCreationTime',
   // SubSecTimeOriginal: 'SubSecTimeOriginal',
   // SubSecTimeDigitized: 'SubSecTimeDigitized',
   // Artist: 'Artist',
   // ByLine: 'By-line',
   // Creator: 'Creator',
   // Copyright: 'Copyright',
   // CopyrightFlag: 'CopyrightFlag',
   // CopyrightNotice: 'CopyrightNotice',
   // Rights: 'Rights',
   // Marked: 'Marked',
   // Title: 'Title',
   // Subject: 'Subject',
   // CaptionAbstract: 'Caption-Abstract',
   // ExposureTime: 'ExposureTime',
   // ExposureMode: 'ExposureMode',
   // ExposureProgram: 'ExposureProgram',
   // ExposureCompensation: 'ExposureCompensation',
   // WhiteBalance: 'WhiteBalance',
   // FocalLength: 'FocalLength',
   // FocalLengthIn35mmFormat: 'FocalLengthIn35mmFormat',
   // ApproximateFocusDistance: 'ApproximateFocusDistance',
   // Aperture: 'FNumber',
   // MaxAperture: 'MaxApertureValue',
   // ISO: 'ISO',
   // MeteringMode: 'MeteringMode',
   // SensitivityType: 'SensitivityType',
   // SensingMethod: 'SensingMethod',
   // ExifVersion: 'ExifVersion',
   // XmpToolkit: 'XMPToolkit',
   // LightSource: 'LightSource',
   // Flash: 'Flash',
   // FileSource: 'FileSource',
   // SceneType: 'SceneType',
   // SceneCaptureType: 'SceneCaptureType',
   // CustomRendered: 'CustomRendered',
   // DigitalZoomRatio: 'DigitalZoomRatio',
   // GainControl: 'GainControl',
   // Contrast: 'Contrast',
   // Saturation: 'Saturation',
   // Sharpness: 'Sharpness',
   // ColorTransform: 'ColorTransform',
   // Compression: 'Compression',
   // Format: 'Format',
   // SubjectDistance: 'SubjectDistanceRange',
   // GpsVersionID: 'GPSVersionID',
   // GpsLatitudeRef: 'GPSLatitudeRef',
   // GpsLatitude: 'GPSLatitude',
   // GpsLongitudeRef: 'GPSLongitudeRef',
   // GpsLongitude: 'GPSLongitude',
   // ThumbnailOffset: 'ThumbnailOffset',
   // ThumbnailLength: 'ThumbnailLength',
   // PhotoshopThumbnail: 'PhotoshopThumbnail',
   // IptcDigest: 'IPTCDigest',
   // DctEncodeVersion: 'DCTEncodeVersion',
   // ImageNumber: 'ImageNumber',
   // DocumentID: 'DocumentID',
   // OriginalDocumentID: 'OriginalDocumentID',
   // DerivedFromDocumentID: 'DerivedFromDocumentID',
   // DerivedFromOriginalDocumentID: 'DerivedFromOriginalDocumentID',
   // InstanceID: 'InstanceID',
   // CodedCharacterSet: 'CodedCharacterSet',
   // ObjectName: 'ObjectName',
   // Keywords: 'Keywords',
   // City: 'City',
   // Location: 'Location',
   // SubLocation: 'Sub-location',
   // State: 'Province-State',
   // Country: 'Country-PrimaryLocationName',
   // ViewingIlluminant: 'ViewingCondIlluminant',
   // ViewingSurround: 'ViewingCondSurround',
   // ViewingIlluminantType: 'ViewingCondIlluminantType',
   // MeasurementObserver: 'MeasurementObserver',
   // MeasurementBacking: 'MeasurementBacking',
   // MeasurementGeometry: 'MeasurementGeometry',
   // MeasurementFlare: 'MeasurementFlare',
   // MeasurementIlluminant: 'MeasurementIlluminant',
   // HistoryAction: 'HistoryAction',
   // HistoryParameters: 'HistoryParameters',
   // HistoryInstanceID: 'HistoryInstanceID',
   // HistoryWhen: 'HistoryWhen',
   // HistorySoftware: 'HistorySoftwareAgent',
   // HistoryChanged: 'HistoryChanged'
   // }
}
