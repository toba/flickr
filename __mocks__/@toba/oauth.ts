export class Client {
   urls: { [key: string]: string };

   constructor(
      requestTokenUrl: string,
      accessTokenUrl: string,
      apiKey: string,
      secret: string,
      version: string,
      callbackUrl: string,
      hashing: string
   ) {
      this.urls = {
         requestTokenUrl,
         accessTokenUrl,
         callbackUrl
      };
   }
}
