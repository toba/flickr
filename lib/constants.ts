export enum Url {
   Host = 'api.flickr.com',
   Base = '/services/rest/',
   RequestToken = 'http://www.flickr.com/services/oauth/request_token',
   Authorize = 'http://www.flickr.com/services/oauth/authorize',
   AccessToken = 'http://www.flickr.com/services/oauth/access_token',
   PhotoSet = 'http://www.flickr.com/photos/trailimage/sets/'
}

enum PhotoMethods {
   EXIF = 'photos.getExif',
   Search = 'photos.search',
   Sets = 'photos.getAllContexts',
   Sizes = 'photos.getSizes',
   Tags = 'tags.getListUserRaw'
}

enum SetMethods {
   Info = 'photosets.getInfo',
   Photos = 'photosets.getPhotos'
}

export const Method = {
   Collections: 'collections.getTree',
   Photo: PhotoMethods,
   Set: SetMethods
};
