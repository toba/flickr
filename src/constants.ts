export const enum Url {
   Host = 'api.flickr.com',
   Base = '/services/rest/',
   RequestToken = 'http://www.flickr.com/services/oauth/request_token',
   Authorize = 'http://www.flickr.com/services/oauth/authorize',
   AccessToken = 'http://www.flickr.com/services/oauth/access_token',
   PhotoSet = 'http://www.flickr.com/photos/trailimage/sets/'
}

enum PhotoMethods {
   EXIF = 'photos.getExif',
   Info = 'photos.getInfo',
   Search = 'photos.search',
   Sets = 'photos.getAllContexts',
   Sizes = 'photos.getSizes',
   Tags = 'tags.getListUserRaw'
}

enum SetMethods {
   Info = 'photosets.getInfo',
   Photos = 'photosets.getPhotos'
}

export const Method = {
   Prefix: 'flickr.',
   Collections: 'collections.getTree',
   Photo: PhotoMethods,
   Set: SetMethods
};
