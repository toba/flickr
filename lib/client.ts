import { Flickr } from './types';
import { Client as AuthClient, Config as AuthConfig } from '@toba/oauth';
import { is, merge } from '@toba/tools';
import { Url, Method } from './constants';
import { call, Identity, Request } from './api';

export interface FeatureSet {
   id: string;
   title: string;
}

export interface ClientConfig {
   userID: string;
   appID: string;
   /** Whether to cache API resuts */
   useCache?: boolean;
   /** Maximum number of responses to cache */
   maxCacheSize?: number;
   /** Optional set IDs to feature */
   featureSets?: FeatureSet[];
   /** Optional set IDs to exclude from results */
   excludeSets?: string[];
   /** Optional tags to exclude from tag request */
   excludeTags?: string[];
   /** Photo sizes to return from search request */
   searchPhotoSizes?: Flickr.SizeUrl[];
   /** Photo sizes to return for photo set request */
   setPhotoSizes?: Flickr.SizeUrl[];
   /** Number of times to retry failed requests */
   maxRetries?: number;
   /** Milliseconds to wait before retrying failed request */
   retryDelay?: number;
   /** https://www.flickr.com/services/api/auth.oauth.html */
   auth: AuthConfig;
}

const defaultConfig: ClientConfig = {
   userID: null,
   appID: null,
   useCache: true,
   maxCacheSize: 200,
   featureSets: [],
   excludeSets: [],
   excludeTags: [],
   searchPhotoSizes: [Flickr.SizeUrl.Large1024],
   setPhotoSizes: [Flickr.SizeUrl.Large1024],
   maxRetries: 3,
   retryDelay: 500,
   auth: null
};

export class FlickrClient {
   config: ClientConfig;
   oauth: AuthClient;

   constructor(config: ClientConfig) {
      this.config = merge(defaultConfig, config);
      this.oauth = new AuthClient(
         Url.RequestToken,
         Url.AccessToken,
         config.auth.apiKey,
         config.auth.secret,
         '1.0A',
         config.auth.callback,
         'HMAC-SHA1'
      );
   }

   get _userID(): Identity {
      return { type: Flickr.TypeName.User, value: this.config.userID };
   }

   _setID(id: string): Identity {
      return { type: Flickr.TypeName.Set, value: id };
   }

   _photoID(id: string | number): Identity {
      return { type: Flickr.TypeName.Photo, value: id.toString() };
   }

   _api<T>(method: string, id: Identity, req: Request<T>): Promise<T> {
      req.auth = this.oauth;
      return call<T>(method, id, req, this.config);
   }

   /**
    * https://www.flickr.com/services/api/flickr.collections.getTree.html
    */
   getCollections() {
      return this._api<Flickr.Collection[]>(Method.Collections, this._userID, {
         parse: r => r.collections.collection,
         allowCache: true
      });
   }

   /**
    * https://www.flickr.com/services/api/flickr.photosets.getInfo.html
    */
   getSetInfo(id: string) {
      return this._api<Flickr.SetInfo>(Method.Set.Info, this._setID(id), {
         parse: r => r.photoset as Flickr.SetInfo,
         allowCache: true
      });
   }

   /**
    * https://www.flickr.com/services/api/flickr.photos.getSizes.html
    */
   getPhotoSizes(id: string) {
      return this._api<Flickr.Size[]>(Method.Photo.Sizes, this._photoID(id), {
         parse: r => r.sizes.size
      });
   }

   /**
    * All sets that a photo belongs to.
    * https://www.flickr.com/services/api/flickr.photos.getAllContexts.html
    */
   getPhotoContext(id: string) {
      return this._api<Flickr.MemberSet[]>(
         Method.Photo.Sets,
         this._photoID(id),
         {
            parse: r => r.set
         }
      );
   }

   /**
    * https://www.flickr.com/services/api/flickr.photos.getExif.html
    */
   getExif(id: string) {
      return this._api<Flickr.Exif[]>(Method.Photo.EXIF, this._photoID(id), {
         parse: r => r.photo.exif,
         allowCache: true
      });
   }

   /**
    * All photos in a set.
    * https://www.flickr.com/services/api/flickr.photosets.getPhotos.html
    */
   getSetPhotos(id: string) {
      return this._api<Flickr.SetPhotos>(Method.Set.Photos, this._setID(id), {
         params: {
            extras:
               [
                  Flickr.Extra.Description,
                  Flickr.Extra.Tags,
                  Flickr.Extra.DateTaken,
                  Flickr.Extra.Location,
                  Flickr.Extra.PathAlias
               ].join() +
               ',' +
               this.config.searchPhotoSizes.join()
         },
         parse: r => r.photoset as Flickr.SetPhotos,
         allowCache: true
      });
   }

   /**
    * The documentation says signing is not required but results differ even
    * with entirely public photos -- perhaps a Flickr bug
    *
    * https://www.flickr.com/services/api/flickr.photos.search.html
    */
   photoSearch(tags: string | string[]) {
      return this._api<Flickr.PhotoSummary[]>(
         Method.Photo.Search,
         this._userID,
         {
            params: {
               extras: this.config.searchPhotoSizes.join(),
               tags: is.array(tags) ? tags.join() : tags,
               sort: Flickr.Sort.Relevance,
               per_page: 500 // maximum
            },
            parse: r => r.photos.photo as Flickr.PhotoSummary[],
            sign: true
         }
      );
   }

   /**
    * All photo tags for API user.
    * https://www.flickr.com/services/api/flickr.tags.getListUserRaw.html
    */
   getAllPhotoTags() {
      return this._api<Flickr.Tag[]>(Method.Photo.Tags, this._userID, {
         parse: r => r.who.tags.tag,
         sign: true,
         allowCache: true
      });
   }
}
