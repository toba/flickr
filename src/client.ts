import { AuthClient, SigningMethod, Token } from '@toba/oauth';
import { is, merge } from '@toba/node-tools';
import { Flickr } from './types';
import { ClientConfig, defaultConfig } from './config';
import { ChangeSubscription, Changes, EventType } from './subscription';
import { Url, Method } from './constants';
import { call, Identity, Request } from './api';
import { cache } from './cache';

export class FlickrClient {
   private config: ClientConfig;
   private oauth: AuthClient;
   subscription: ChangeSubscription;

   constructor(config: ClientConfig) {
      this.config = merge(defaultConfig, config);
      this.subscription = new ChangeSubscription(this);
      this.oauth = new AuthClient(
         Url.RequestToken,
         Url.AccessToken,
         config.auth.apiKey,
         config.auth.secret,
         '1.0A',
         config.auth.callback,
         SigningMethod.HMAC
      );
      cache.maxItems(this.config.maxCacheSize);

      if (this.config.useCache) {
         // subscribe to change events to clear cache if an external watcher is
         // added
         this.subscription.addEventListener(EventType.NewWatcher, () => {
            this.subscription.addEventListener(
               EventType.Change,
               this.onChange.bind(this)
            );
         });
      }
   }

   private get userID(): Identity {
      return { type: Flickr.TypeName.User, value: this.config.userID };
   }

   private setID = (id: string): Identity => ({
      type: Flickr.TypeName.Set,
      value: id
   });

   private photoID = (id: string | number): Identity => ({
      type: Flickr.TypeName.Photo,
      value: id.toString()
   });

   private apiCall<T>(
      method: string,
      id: Identity,
      req: Request<T>
   ): Promise<T | null> {
      req.auth = this.oauth;
      return call<T>(method, id, req, this.config);
   }

   /**
    * Remove items from cache when change is detected on the Flickr server.
    */
   private onChange(changes: Changes) {
      if (!this.config.useCache) {
         return;
      }

      if (changes.collections.length > 0) {
         cache.remove(Method.Collections, this.userID.value);
      }

      changes.sets.forEach(id => {
         cache.remove(Method.Set.Info, id);
         cache.remove(Method.Set.Photos, id);
      });
   }

   /**
    * Add method to receive change notifications. This also has the effect of
    * activating change polling.
    */
   subscribe(fn: (change: Changes) => void) {
      this.subscription.add(fn);
   }

   /**
    * @see https://www.flickr.com/services/api/flickr.collections.getTree.html
    */
   async getCollections(allowCache = true) {
      const collections = await this.apiCall<Flickr.Collection[]>(
         Method.Collections,
         this.userID,
         {
            select: r =>
               r.collections !== undefined ? r.collections.collection : [],
            allowCache
         }
      );
      if (collections !== null) {
         this.subscription.updateCollections(...collections);
      }
      return collections;
   }

   /**
    * @see https://www.flickr.com/services/api/flickr.photosets.getInfo.html
    */
   async getSetInfo(
      id: string,
      allowCache = true
   ): Promise<Flickr.SetInfo | null> {
      const info = await this.apiCall<Flickr.SetInfo>(
         Method.Set.Info,
         this.setID(id),
         {
            select: r => r.photoset as Flickr.SetInfo,
            allowCache
         }
      );
      if (info !== null) {
         this.subscription.updateSet(info.id, parseInt(info.date_update));
      }
      return info;
   }

   /**
    * All photos in a set. Include last update time to enable change detection.
    * @see https://www.flickr.com/services/api/flickr.photosets.getPhotos.html
    */
   async getSetPhotos(
      id: string,
      extras: Flickr.Extra[] = [],
      allowCache = true
   ): Promise<Flickr.SetPhotos | null> {
      const extrasList =
         extras.length > 0
            ? extras.join()
            : [
                 Flickr.Extra.Description,
                 Flickr.Extra.Tags,
                 Flickr.Extra.DateTaken,
                 Flickr.Extra.DateUpdated,
                 Flickr.Extra.Location,
                 Flickr.Extra.PathAlias
              ].join() +
              ',' +
              this.config.setPhotoSizes.join();

      const photos = await this.apiCall<Flickr.SetPhotos>(
         Method.Set.Photos,
         this.setID(id),
         {
            params: {
               extras: extrasList
            },
            select: r => r.photoset as Flickr.SetPhotos,
            allowCache
         }
      );
      if (photos !== null) {
         this.subscription.updateSet(id, photos);
      }
      return photos;
   }

   /**
    * @see https://www.flickr.com/services/api/flickr.photos.getInfo.html
    */
   getPhotoInfo = (id: string) =>
      this.apiCall<Flickr.PhotoInfo>(Method.Photo.Info, this.photoID(id), {
         select: r => r.photo as Flickr.PhotoInfo,
         allowCache: true
      });

   /**
    * @see https://www.flickr.com/services/api/flickr.photos.getSizes.html
    */
   getPhotoSizes = (id: string) =>
      this.apiCall<Flickr.Size[]>(Method.Photo.Sizes, this.photoID(id), {
         select: r => (r.sizes !== undefined ? r.sizes.size : [])
      });

   /**
    * All sets that a photo belongs to.
    * @see https://www.flickr.com/services/api/flickr.photos.getAllContexts.html
    */
   getPhotoContext = (id: string) =>
      this.apiCall<Flickr.MemberSet[]>(Method.Photo.Sets, this.photoID(id), {
         select: r => (r.set !== undefined ? r.set : [])
      });

   /**
    * @see https://www.flickr.com/services/api/flickr.photos.getExif.html
    */
   getExif = (id: string) =>
      this.apiCall<Flickr.Exif[]>(Method.Photo.EXIF, this.photoID(id), {
         select: r => {
            if (r.photo === undefined) {
               return [];
            }
            if (r.photo.exif !== undefined) {
               return r.photo.exif;
            }
            // Flickr changed the field name
            if (r.photo.EXIF !== undefined) {
               return r.photo.EXIF;
            }
            return [];
         },
         allowCache: true
      });

   /**
    * The documentation says signing is not required but results differ even
    * with entirely public photos -- perhaps a Flickr bug
    *
    * @see https://www.flickr.com/services/api/flickr.photos.search.html
    */
   photoSearch = (tags: string | string[]) =>
      this.apiCall<Flickr.PhotoSummary[]>(Method.Photo.Search, this.userID, {
         params: {
            extras: this.config.searchPhotoSizes.join(),
            tags: is.array(tags) ? tags.join() : tags,
            sort: Flickr.Sort.Relevance,
            per_page: 500 // maximum
         },
         select: r =>
            r.photos !== undefined
               ? (r.photos.photo as Flickr.PhotoSummary[])
               : [],
         sign: true
      });

   /**
    * All photo tags for API user.
    * @see https://www.flickr.com/services/api/flickr.tags.getListUserRaw.html
    */
   getAllPhotoTags = () =>
      this.apiCall<Flickr.Tag[]>(Method.Photo.Tags, this.userID, {
         select: r => (r.who !== undefined ? r.who.tags.tag : []),
         sign: true,
         allowCache: true
      });

   getRequestToken = (): Promise<string> =>
      new Promise<string>((resolve, reject) => {
         const token = this.config.auth.token;
         if (token === undefined) {
            return reject('Cannot get request token without OAuth object');
         }
         this.oauth.getOAuthRequestToken((error, requestToken, secret) => {
            if (is.value(error)) {
               return reject(error);
            }
            // token and secret are both needed for the next call but token is
            // echoed back from the authorize service
            token.request = requestToken;
            token.secret = secret;

            resolve(`${Url.Authorize}?oauth_token=${requestToken}`);
         });
      });

   getAccessToken = (requestToken: string, verifier: string) =>
      new Promise<Token>((resolve, reject) => {
         const token = this.config.auth.token;
         if (token === undefined || token.secret === undefined) {
            return reject('Cannot get access token without secret');
         }

         this.oauth.getOAuthAccessToken(
            requestToken,
            token.secret,
            verifier,
            (error, accessToken, secret) => {
               token.secret = undefined;
               if (is.value(error)) {
                  return reject(error);
               }
               resolve({
                  access: accessToken,
                  secret: secret
               } as Token);
            }
         );
      });

   /**
    * Empty API response cache.
    */
   clearCache() {
      cache.clear();
   }
}
