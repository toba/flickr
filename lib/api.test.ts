import {
   call,
   parse,
   retry,
   parameterize,
   failResponse,
   basicGet,
   Request,
   ResponseHandler,
   Identity
} from './api';
import { config } from './client.test';
import { Url, Method } from './constants';
import { Flickr } from './types';
import { merge } from '@toba/utility';
import { log } from '@toba/logger';

const key = 'mockKey';
const logMock = jest.fn();
const responseHandler = jest.fn();
const mockID: Identity = { value: 'user-name', type: Flickr.TypeName.User };
const mockRequest: Request<Flickr.Collection[]> = {
   parse: r => r.collections.collection,
   allowCache: false
};
let logWithColor: boolean;

console.log = logMock;

// remove color codes since they complicate the snapshots
beforeAll(() => {
   logWithColor = log.config.color;
   log.update({ color: false });
});
beforeEach(() => logMock.mockClear());
afterAll(() => {
   log.update({ color: logWithColor });
});

test('builds request parameters', () => {
   const url = parameterize(Method.Collections, mockID, mockRequest, config);

   expect(url).toBe(
      `?api_key=${config.auth.apiKey}&format=json&nojsoncallback=1&method=${
         Method.Prefix
      }${Method.Collections}&user_id=user-name`
   );
});

test('curries basic HTTP get method', () => {
   const url = parameterize(Method.Collections, mockID, mockRequest, config);
   const getter = basicGet(url, responseHandler);
   expect(getter).toBeInstanceOf(Function);

   return getter().then(() => {
      expect(responseHandler).toHaveBeenCalledTimes(1);
   });
});

test('parses response body', () => {
   const notFound: Flickr.Response = {
      stat: Flickr.Status.Failed,
      message: 'thing not found'
   };
   // empty body should fail with retry option
   expect(parse(null, key)).toEqual(failResponse);
   // html body should also fail
   expect(parse('<html>whatever</html>', key)).toEqual(failResponse);
   // non-JSON should fail
   expect(parse('not json', key)).toEqual(failResponse);
   // Flickr IDs that can't be found shouldn't be retried
   expect(parse(JSON.stringify(notFound), key)).toEqual(
      merge(notFound, { retry: false })
   );

   expect(logMock).toMatchSnapshot();
});

test('retries request', () => {
   const mockCall = jest
      .fn(() => Promise.resolve())
      .mockImplementationOnce(() => Promise.reject('reason 1'))
      .mockImplementationOnce(() => Promise.reject('reason 2'));

   jest.useFakeTimers();

   // do not retry if count >= max retries
   config.maxRetries = 0;
   expect(retry(mockCall, key, config)).toBe(false);
   expect(mockCall).toHaveBeenCalledTimes(0);

   config.maxRetries = 2;
   expect(retry(mockCall, key, config)).toBe(true);
   expect(setTimeout).toHaveBeenCalledTimes(1);

   // jest.runAllTimers();
   // expect(mockCall).toHaveBeenCalledTimes(1);
});

test('converts response to objects', () =>
   call(
      Method.Collections,
      { type: Flickr.TypeName.User, value: '' },
      { parse: r => r },
      config
   ).then((res: Flickr.Response) => {
      expect(res).toHaveProperty('stat', Flickr.Status.Okay);
      expect(res).toHaveProperty('collections');
      expect(res.collections).toHaveProperty('collection');
      expect(res.collections.collection).toBeInstanceOf(Array);
   }));
