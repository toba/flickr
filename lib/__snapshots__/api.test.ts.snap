// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses response body 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[Error] Call to mockKey returned null key=mockKey level=4 message=\\"Call to mockKey returned null\\"",
    ],
    Array [
      "[Error] Flickr returned HTML instead of JSON key=mockKey level=4 message=\\"Flickr returned HTML instead of JSON\\"",
    ],
    Array [
      "[Error] Unexpected token o in JSON at position 1 key=mockKey error=\\"SyntaxError: Unexpected token o in JSON at position 1\\" stack=\\"SyntaxError: Unexpected token o in JSON at position 1
    at JSON.parse (<anonymous>)
    at Object.parse (D:\\\\\\\\dev\\\\\\\\src\\\\\\\\github.com\\\\\\\\toba\\\\\\\\flickr\\\\\\\\lib\\\\\\\\api.ts:199:18)
    at Object.<anonymous>.test (D:\\\\\\\\dev\\\\\\\\src\\\\\\\\github.com\\\\\\\\toba\\\\\\\\flickr\\\\\\\\lib\\\\\\\\api.test.ts:70:11)
    at Object.asyncFn (D:\\\\\\\\dev\\\\\\\\src\\\\\\\\github.com\\\\\\\\toba\\\\\\\\flickr\\\\\\\\node_modules\\\\\\\\jest-jasmine2\\\\\\\\build\\\\\\\\jasmine_async.js:82:37)
    at resolve (D:\\\\\\\\dev\\\\\\\\src\\\\\\\\github.com\\\\\\\\toba\\\\\\\\flickr\\\\\\\\node_modules\\\\\\\\jest-jasmine2\\\\\\\\build\\\\\\\\queue_runner.js:52:12)
    at new Promise (<anonymous>)
    at mapper (D:\\\\\\\\dev\\\\\\\\src\\\\\\\\github.com\\\\\\\\toba\\\\\\\\flickr\\\\\\\\node_modules\\\\\\\\jest-jasmine2\\\\\\\\build\\\\\\\\queue_runner.js:39:19)
    at promise.then (D:\\\\\\\\dev\\\\\\\\src\\\\\\\\github.com\\\\\\\\toba\\\\\\\\flickr\\\\\\\\node_modules\\\\\\\\jest-jasmine2\\\\\\\\build\\\\\\\\queue_runner.js:73:82)
    at <anonymous>\\" level=4 message=\\"Unexpected token o in JSON at position 1\\"",
    ],
    Array [
      "[Error] thing not found key=mockKey code= level=4 message=\\"thing not found\\"",
    ],
  ],
}
`;
